docs/proposals/20191030-machine-health-checking.md

```diff
@@ -0,0 +1,214 @@
+        - [Beta -> GA Graduation](#beta---ga-graduation)
+        - [Removing a deprecated flag](#removing-a-deprecated-flag)
+    - [Upgrade / Downgrade Strategy](#upgrade--downgrade-strategy)
+    - [Version Skew Strategy](#version-skew-strategy)
+  - [Implementation History](#implementation-history)
+  - [Drawbacks [optional]](#drawbacks-optional)
+  - [Alternatives [optional]](#alternatives-optional)
+  - [Infrastructure Needed [optional]](#infrastructure-needed-optional)
+
+## Glossary
+Refer to the [Cluster API Book Glossary](https://cluster-api.sigs.k8s.io/reference/glossary.html).
+
+## Summary
+Enable opt in automated health checking and remediation of unhealthy nodes backed by machines.
+
+## Motivation
+- Reduce administrative overhead to run a cluster.
+
+- Ability to respond to failures of machines and keep the cluster nodes healthy.
+
+### Goals
+- Enable automated remediation for pools of machines/nodes (e.g a machineSet)
+
+- Allow to define different unhealthy criterias based on node conditions for different pools.
+
+- Providing administrator-defined short-circuiting of automated remediation when multiple nodes are unhealthy at the same time.
+
+### Non-Goals/Future Work
+- Recording long-term stable history of all health-check failures or remediations.
+
+- Providing a mechanisim to guarantee that application quorum for N members is maintained at any time.
+
+## Proposal
+The machine health checker (MHC) does a best effort to keep nodes healthy in the cluster, by removing machines backing unhealthy Nodes.
+
+It provides a short-circuit mechanism and limits remediation when the `maxUnhealthy` threshold is reached for a targeted group of machines. This is similar to what the node life cycle controller does for reducing the eviction rate as nodes become unhealthy in a given zone. E.g a large number of nodes in a single zone are down due to a networking issue.
+
+The machine health checker is an integration point between node problem detection tooling expresed as node conditions and remediation to achieve a node auto repairing feature.
+
+### Unhealthy criteria:
+A machine is unhealthy when:
+
+- The referenced node meets the unhealthy node conditions criteria defined.
+- The Machine has no nodeRef.
+- The Machine has a nodeRef but the referenced node is not found.
+- The Machine is in phase "Failed", It has `ErrorMessage`/`ErrorReason` set.
+
+If any of those criteria are met for longer than the given timeouts and the `maxUnhealthy` threshold has not been reached yet, remediation is triggered.
+For the node conditions the time outs are defined by the admin. For the other cases opinionated values can be assumed.
+For a machine with no nodeRef an opinionated value could be assumed e.g 10 min.
+For a node notFound or a failed machine, the machine is considerable unrecoverable, remediation can be triggered right away.
+
+### Remediation:
+- A deletion request for the machine is sent to the API server.
```

Comment:
Should this not allow for a `maxSurge` like the RollingUpdate MachineDeploymentStrategyType?

If remediation does not use or coordinate with the MachineDeployment, then a concurrent RollingUpdate and remediation could end up removing more nodes than the application can tolerate.

Reply:
To expand on this, I worry that cluster-api is walking into a failed design. If there are numerous subsystems each trying to drain and/or delete nodes independently from each other, it is likely that they will occasionally overstress applications, causing databases to drop below quorum and the like.

There should be one controller responsible for doing rolling updates of nodes. Other subsystems could then independently nominate/mark nodes for the rolling update controller to get rid of.

Reply:
This is a good point. This proposal is actually trying to respect the right boundaries/responsibilities between controllers and using the right consumable semantics that allow composability to manipulate machine resources and signal other controllers.
The MHC is just nominating a machine for deletion by signalling the api server a desire for a machine resource to be deleted; this could also be requested anytime by e.g a user. As soon as the deletionTimestamp is set any watcher can react as they see fit.
The replicas reconciliation is so delegated to where it belongs, the controller owning the machine e.g machineDeployment, machineSet, other...
The actual deletion process is so delegated to where it belongs, the machine controller, and this is the layer where the application disruption toleration must be enforced via draining and PDB.

Reply:
It is not clear to me that this mechanism would respect the `maxSurge` and `maxUnavailable` settings of the machineDeployment. I believe the MDC should use a mechanism that respects those settings, especially if the machine has workloads on it that may still be functioning.

This might be more of a machineDeployment/machineSet issue, but I can see a distinction between requesting voluntary and involuntary eviction of workloads. Deletion seems like it would tend towards the involuntary side of things.

Reply:
```suggestion
- The MDC requests deletion of the machine by placing a `cluster.k8s.io/delete-machine` annotation on it.
```

Reply:
There's no such annotation. The process is the same as for any other entity signalling a deletion desire. The machine health check sends a request to the API server i.e `client.delete()`. Everything else happens out of band:
- The machine is give a deletionTimestamp.
- Any watcher can adjust based on the deletionTimestamp (usually by filtering)
- The machine controller sees the deletionTimestamp and enforces workloads availability policy by draining and pod pdb.
- The machine controller removes the machine finalizer.
- The api server removes the machine object from etcd.
You can also still set the maxUnhealthy machine health checker field to short circuit and to be less tolerant than the deployment maxUnhealthy.

Reply:
There is such an annotation, at https://github.com/kubernetes-sigs/cluster-api/blob/aa3d2c200147e18e1bea289b428037d0c91bcda0/controllers/machineset_delete_policy.go#L36

Reply:
Ah got you, that's for then machineSet to prioritise deletion during a scale down operation. A machine with a deletionTimestamp won't even make it to the list of machines being prioritised, It will be filtered out and dismissed when reconciling towards expected number of replicas